File: ruby/largest_rectangle.rb
# https://leetcode.com/problems/largest-rectangle-in-histogram/description/

# @param {Integer[]} heights
# @return {Integer}
def largest_rectangle_area(heights)
  return 0 if heights.size == 0
  rects = []
  heights.each_with_index do |height, index|
    forward, backward = 0, 0
    while(index + forward < heights.size)
      break if heights[index + forward] < height
      forward += 1
    end

    while(index - backward >= 0)
      break if heights[index - backward] < height
      backward += 1
    end
      
    rects << height * (forward + backward - 1)
  end

  rects.max
end

File: go/largest_rectangle.go
func largestRectangleArea(heights []int) int {
  if len(heights) == 0 {
    return 0
  }

  if len(heights) == 1 {
    return heights[0]
  }

  var rects []int

  iter := 0
  for iter < len(heights) {
    height := heights[iter]

    forward, backward := 0, 0
    for iter + forward < len(heights) {
      if heights[iter + forward] < height { break } 
      forward ++
    }

    for iter - backward >= 0 {
      if heights[iter - backward] < height { break }
      backward ++
    }
      
    iter ++

    value := height * (forward + backward - 1)
    rects = append(rects, value)
  }

  max := rects[0]
  for _ , value := range(rects) {
    if value > max {
      max = value
    }
  }

  return max
}